### PR TITLE
feat: delete Tensorboards with delete API [DET-5143]

### PR DESCRIPTION
**Pull request by stoksc**
## Description
This change adds code to GC Tensorboards when DELETE experiment is run. It is piggybacked as a new flag onto the checkpoint GC script.

## Test Plan
- [x] test with shared fs
- [x] test with s3
- [ ] test with hdfs
- [x] test with gcs bucket

## Commentary (optional)
Looking for a bit of feedback on whether the interface is OK.. I usually don't touch this harness stuff.

[DET-1234]: https://determinedai.atlassian.net/browse/DET-1234

### CODE REVIEW

**Comment by stoksc** [comment]
blocked on #2056 but if you want to 👁️ the last commit @rb-determined-ai, feel free.

**Comment by shiyuann** [comment]
@stoksc is this PR for deleting experiments or just TensorBoards?

**Comment by stoksc** [comment]
@shiyuann yeah it is, but it's blocked on another.

File: harness/determined/tensorboard/s3.py [hunk 1]
```diff
@@ -40,3 +46,13 @@ def sync(self) -> None:
 
             self.client.upload_file(str(path), self.bucket, key_name)
             self._synced_event_sizes[path] = path.stat().st_size
+
+    def delete(self) -> None:
+        self.resource.Bucket(self.bucket).objects.filter(Prefix=str(self.sync_path)).delete()
+
+        # If we're attached to using the low level APIs:
```
Comment (stoksc) [inline harness/determined/tensorboard/s3.py:53]: preference @rb-determined-ai ?

Comment (rb-determined-ai) [reply harness/determined/tensorboard/s3.py:53]: The StorageManager code is some of the oldest code we have.  I don't think we have any attachment to the low-level APIs, that's just how it's always been.

Comment (rb-determined-ai) [reply harness/determined/tensorboard/s3.py:53]: (so I prefer the easy way)

File: harness/determined/exec/gc_checkpoints.py [hunk 2]
```diff
@@ -46,6 +57,7 @@ def main(argv: List[str]) -> None:
         action="version",
         version="Determined checkpoint GC, version {}".format(det.__version__),
     )
+    parser.add_argument("--experiment-id", help="The experiment ID to run the GC job for")
```
Comment (rb-determined-ai) [inline harness/determined/exec/gc_checkpoints.py:60]: It occurs to me that some day when we want to version the harness/master protocol, we'll have to either pack the gc script in at runtime or we will have to edit it several versions before we start using a new feature in it.

Just a thought, not related to this PR in particular.

File: harness/determined/exec/gc_checkpoints.py [hunk 3]
```diff
@@ -64,6 +76,12 @@ def main(argv: List[str]) -> None:
         default=os.getenv("DET_DELETE", []),
         help="Checkpoints to delete (JSON-formatted file)",
     )
+    parser.add_argument(
+        "--delete-tensorboards",
+        action="store_true",
+        default=False,
```
Comment (rb-determined-ai) [inline harness/determined/exec/gc_checkpoints.py:82]: maybe accept the environment variable to be consistent?  Or delete all the env variables that we haven't used in like forever...

Either way.

File: harness/determined/tensorboard/base.py [hunk 4]
```diff
@@ -14,7 +14,11 @@ class TensorboardManager:
     implements the sync method.
     """
 
-    def __init__(self, base_path: pathlib.Path, sync_path: pathlib.Path):
+    def __init__(
+        self,
+        base_path: pathlib.Path,
+        sync_path: pathlib.Path,
+    ):
```
Comment (rb-determined-ai) [inline harness/determined/tensorboard/base.py:21]: wait... this file isn't being linted, or this `__init__` without a `-> None` would throw an error... right?

File: harness/determined/tensorboard/base.py [hunk 5]
```diff
@@ -53,3 +57,9 @@ def sync(self) -> None:
         Save the object to the backing persistent storage.
         """
         pass
+
+    def delete(self) -> None:
+        """
+        Delete all objects from the backing persistent storage.
+        """
+        pass
```
Comment (rb-determined-ai) [inline harness/determined/tensorboard/base.py:65]: If this method isn't optional (seems like it isn't), we should either raise a NotImplementedError or (better) we should do `class TensorboardManager(metaclass=abc.ABCmeta` and decorate `delete` with `@abc.abstractmethod`

File: harness/determined/tensorboard/s3.py (hunk 1 above)
Comment (stoksc) [reply harness/determined/tensorboard/s3.py:53]: I wasn't sure if it was some weird minio compat thing I didn't know about, but minio looked like it should work with both APIs from what I saw (I'm unsure if it's actually a different API or it's just a nice client side wrapper around the low levels). I'll just use the nice one.

**Review by rb-determined-ai — COMMENTED** [review]
Please look into the linting thing, but otherwise this looks great!

File: harness/determined/tensorboard/s3.py (hunk 1 above)
Comment (rb-determined-ai) [reply harness/determined/tensorboard/s3.py:53]: nah, our s3 code was written way before we ran it against minio

File: harness/determined/tensorboard/base.py (hunk 4 above)
Comment (stoksc) [reply harness/determined/tensorboard/base.py:21]: i'll peek at this.

File: harness/determined/tensorboard/base.py (hunk 5 above)
Comment (stoksc) [reply harness/determined/tensorboard/base.py:65]: I'll make it an abstract class

File: harness/determined/tensorboard/base.py (hunk 4 above)
Comment (stoksc) [reply harness/determined/tensorboard/base.py]: OK, the file is definitely being linted, e.g. if I remove the ` -> None` from `sync`, it flips out, but if I remove it from any `__init__` call it doesn't seem to mind 🙃 ?

File: harness/determined/exec/gc_checkpoints.py (hunk 2 above)
Comment (stoksc) [reply harness/determined/exec/gc_checkpoints.py:60]: yeah.. that's a good point. interesting.

**Review by rb-determined-ai — APPROVED** [review]

**Pull request merged**